### PR TITLE
Prefer not to spawn within 20 meters of enemy

### PR DESCRIPTION
**Pull request by fvnkhead**
I noticed the `IsSpawnpointValid` function checks for nearby projectiles, but not enemy players. Spawning next to enemies is common enough with a high player count, so this should alleviate the issue. I ran this for about 2 weeks before taking a break from hosting, and during that time, nearby enemy spawns would only occur when running out of valid spawns (eg. on Colony with a high player count, or on LF maps).

Also parameterized the minimum distance into `minEnemyDist`, which is used for checking both projectile and enemy distances, and this has been increased from 600 to 1000 (which is about 20 meters by my calculations?)

Formatting is screwed, I don't know what settings to use for squirrel with vim and they vary across files.

### CODE REVIEW

**Comment by Erlite** [comment]
Pretty sure the projectile check isn't very robust either. I can't tell how many times I've spawned inside a cluster missile or killed people that way.

Awesome PR though!

File: Northstar.CustomServers/mod/scripts/vscripts/mp/spawn.nut [hunk 1]
```diff
@@ -261,10 +261,17 @@ bool function IsSpawnpointValid( entity spawnpoint, int team )
 			return false
 	}
 
-	array<entity> projectiles = GetProjectileArrayEx( "any", TEAM_ANY, TEAM_ANY, spawnpoint.GetOrigin(), 600 )
+    float minEnemyDist = 1000.0 // about 20 meters?
+	array<entity> projectiles = GetProjectileArrayEx( "any", TEAM_ANY, TEAM_ANY, spawnpoint.GetOrigin(), minEnemyDist )
 	foreach ( entity projectile in projectiles )
```
Comment (uniboi) [inline Northstar.CustomServers/mod/scripts/vscripts/mp/spawn.nut:266]:
```suggestion
	foreach ( entity projectile in GetProjectileArrayEx( "any", TEAM_ANY, TEAM_ANY, spawnpoint.GetOrigin() )
```

File: Northstar.CustomServers/mod/scripts/vscripts/mp/spawn.nut [hunk 2]
```diff
@@ -261,10 +261,17 @@ bool function IsSpawnpointValid( entity spawnpoint, int team )
...
 		if ( projectile.GetTeam() != team )
 			return false
+
+    // prefer no enemies nearby
+    array<entity> players = GetPlayerArrayEx( "any", TEAM_ANY, TEAM_ANY, spawnpoint.GetOrigin(), minEnemyDist )
+    foreach ( entity player in players )
```
[the remark this answers is not in the document]
Comment (fvnkhead) [reply Northstar.CustomServers/mod/scripts/vscripts/mp/spawn.nut:272]: they were leftovers from old code, but I'll commit these

**Comment by ASpoonPlaysGames** [comment]
Btw some of the spawn issues that happen are due to that spectator mod, there is a bug with it that spawns you at an enemy team's spawn. Unsure who made the mod though

**Comment by fvnkhead** [comment]
> Btw some of the spawn issues that happen are due to that spectator mod, there is a bug with it that spawns you at an enemy team's spawn. Unsure who made the mod though
I didn't have that issue but I had spectate-after-killcam disabled, whereas V3X had it enabled, and the bad spawns occured there a lot more often.
It's faky's mod I think

**Comment by ASpoonPlaysGames** [comment]
> > Btw some of the spawn issues that happen are due to that spectator mod, there is a bug with it that spawns you at an enemy team's spawn. Unsure who made the mod though
> 
> I didn't have that issue but I had spectate-after-killcam disabled, whereas V3X had it enabled, and the bad spawns occured there a lot more often. It's faky's mod I think

On the V3X server I can consistently spawn in the enemy spawn, the bug is 100% reproduceable, I'll talk to faky

**Comment by Erlite** [comment]
Eh, even on Vanilla servers. An example would by Drydock, the little bridge part where the evac ship can spawn sometimes. I've had people spawn behind me a split second after spawning. 

There really shouldn't be a specific spawn location for teams, it should just rate them all and find the safest one. Maybe go for spawns near teammates that aren't too close to enemies primarily.

File: Northstar.CustomServers/mod/scripts/vscripts/mp/spawn.nut [hunk 3]
```diff
@@ -261,10 +261,15 @@ bool function IsSpawnpointValid( entity spawnpoint, int team )
 			return false
 	}
 
-	array<entity> projectiles = GetProjectileArrayEx( "any", TEAM_ANY, TEAM_ANY, spawnpoint.GetOrigin(), 600 )
-	foreach ( entity projectile in projectiles )
+    const minEnemyDist = 1000.0 // about 20 meters?
+	foreach ( entity projectile in GetProjectileArrayEx( "any", TEAM_ANY, TEAM_ANY, spawnpoint.GetOrigin() )
```
Comment (uniboi) [inline Northstar.CustomServers/mod/scripts/vscripts/mp/spawn.nut:265]:
```suggestion
	foreach ( entity projectile in GetProjectileArrayEx( "any", TEAM_ANY, TEAM_ANY, spawnpoint.GetOrigin(), minEnemyDist )
```
Shouldn't we check for distance to enemy projectiles as well?

**Review by uniboi — CHANGES_REQUESTED** [review]

File: Northstar.CustomServers/mod/scripts/vscripts/mp/spawn.nut (hunk 3 above)
Comment (fvnkhead) [reply Northstar.CustomServers/mod/scripts/vscripts/mp/spawn.nut:265]: Yeah, good find

**Review by uniboi — CHANGES_REQUESTED** [review]
yes

**Review by uniboi — APPROVED** [review]
mfw I clicked the wrong box

File: Northstar.CustomServers/mod/scripts/vscripts/mp/spawn.nut [hunk 4]
```diff
@@ -261,10 +261,15 @@ bool function IsSpawnpointValid( entity spawnpoint, int team )
...
-	foreach ( entity projectile in projectiles )
+    const minEnemyDist = 1000.0 // about 20 meters?
+	foreach ( entity projectile in GetProjectileArrayEx( "any", TEAM_ANY, TEAM_ANY, spawnpoint.GetOrigin(), minEnemyDist )
 		if ( projectile.GetTeam() != team )
 			return false
+
+    // prefer no enemies nearby
+    foreach ( entity player in  GetPlayerArrayEx( "any", TEAM_ANY, TEAM_ANY, spawnpoint.GetOrigin(), minEnemyDist ) )
+        if ( player.GetTeam() != team )
+            return false
```
Comment (uniboi) [inline Northstar.CustomServers/mod/scripts/vscripts/mp/spawn.nut:272]:
```suggestion
    const minEnemyDist = 1000.0 // about 20 meters?
    // in rsquirrel extend returns null unlike in vanilla squirrel
    array<entity> spawnBlockers = GetPlayerArrayEx( "any", TEAM_ANY, TEAM_ANY, spawnpoint.GetOrigin(), minEnemyDist ) )
    spawnBlockers.extend( GetProjectileArrayEx( "any", TEAM_ANY, TEAM_ANY, spawnpoint.GetOrigin(), minEnemyDist ) )
	foreach ( entity blocker in spawnBlockers )
		if ( blocker.GetTeam() != team )
			return false
```

optimization

**Review by uniboi — CHANGES_REQUESTED** [review]

**Review by uniboi — APPROVED** [review]
should also fix spawning near projectiles but I can't test

**Review by F1F7Y — APPROVED** [review]
Code looks good. Didnt test in-game

File: Northstar.CustomServers/mod/scripts/vscripts/mp/spawn.nut [hunk 5]
```diff
@@ -261,9 +261,12 @@ bool function IsSpawnpointValid( entity spawnpoint, int team )
 			return false
 	}
 
-	array<entity> projectiles = GetProjectileArrayEx( "any", TEAM_ANY, TEAM_ANY, spawnpoint.GetOrigin(), 600 )
-	foreach ( entity projectile in projectiles )
-		if ( projectile.GetTeam() != team )
+    const minEnemyDist = 1000.0 // about 20 meters?
+    // in rsquirrel extend returns null unlike in vanilla squirrel
+    array<entity> spawnBlockers = GetPlayerArrayEx( "any", TEAM_ANY, TEAM_ANY, spawnpoint.GetOrigin(), minEnemyDist ) )
```
Comment (tkriik) [inline Northstar.CustomServers/mod/scripts/vscripts/mp/spawn.nut:266]:
```suggestion
    array<entity> spawnBlockers = GetPlayerArrayEx( "any", TEAM_ANY, TEAM_ANY, spawnpoint.GetOrigin(), minEnemyDist )
```

Comment (fvnkhead) [inline Northstar.CustomServers/mod/scripts/vscripts/mp/spawn.nut:266]:
```suggestion
    array<entity> spawnBlockers = GetPlayerArrayEx( "any", TEAM_ANY, TEAM_ANY, spawnpoint.GetOrigin(), minEnemyDist )
```

**Comment by flifloa** [comment]
Some kid on the northstar chat screamed for this months ago and started an argument over the spawn mechanic.

File: Northstar.CustomServers/mod/scripts/vscripts/mp/spawn.nut [hunk 6]
```diff
@@ -261,9 +261,12 @@ bool function IsSpawnpointValid( entity spawnpoint, int team )
 			return false
 	}
 
-	array<entity> projectiles = GetProjectileArrayEx( "any", TEAM_ANY, TEAM_ANY, spawnpoint.GetOrigin(), 600 )
-	foreach ( entity projectile in projectiles )
-		if ( projectile.GetTeam() != team )
+    const minEnemyDist = 1000.0 // about 20 meters?
+    // in rsquirrel extend returns null unlike in vanilla squirrel
+    array<entity> spawnBlockers = GetPlayerArrayEx( "any", TEAM_ANY, TEAM_ANY, spawnpoint.GetOrigin(), minEnemyDist )
```
Comment (ASpoonPlaysGames) [inline Northstar.CustomServers/mod/scripts/vscripts/mp/spawn.nut:266]:
```suggestion
    array< entity > spawnBlockers = GetPlayerArrayEx( "any", TEAM_ANY, TEAM_ANY, spawnpoint.GetOrigin(), minEnemyDist )
```

**Review by ASpoonPlaysGames — APPROVED** [review]
Code looks good, can't hurt

**Pull request merged**